src/kernel/filesystem/fat32.zig

```diff
@@ -0,0 +1,1756 @@
+const std = @import("std");
+const builtin = @import("builtin");
+const expectEqualSlices = std.testing.expectEqualSlices;
+const expectEqual = std.testing.expectEqual;
+const expectError = std.testing.expectError;
+const expect = std.testing.expect;
+const log = std.log.scoped(.fat32);
+const AutoHashMap = std.AutoHashMap;
+const Allocator = std.mem.Allocator;
+const ArrayList = std.ArrayList;
+const vfs = @import("vfs.zig");
+const mem = @import("../mem.zig");
+const CodePage = @import("../code_page.zig").CodePage;
+const mkfat32 = @import("../../../mkfat32.zig");
+
+/// The boot record for FAT32. This is use for parsing the initial boot sector to extract the
+/// relevant information for todays FAT32.
+const BootRecord = struct {
+    /// The jump bytes that begin the boot code. This will jump over the FAT32 header.
+    jmp: [3]u8,
+
+    /// The OEM name.
+    oem: [8]u8,
+
+    /// The number of bytes per sector.
+    bytes_per_sector: u16,
+
+    /// The number of sectors per cluster.
+    sectors_per_cluster: u8,
+
+    /// The number of reserved sectors at the beginning of the image. This is where the fat
+    /// header, FSInfo and FAT is stored.
+    reserved_sectors: u16,
+
+    /// The number of FAT's.
+    fat_count: u8,
+
+    /// The size in bytes of the root directory. This is only used by FAT12 and FAT16, so is
+    /// always 0 for FAT32.
+    zero_root_directory_size: u16,
```

Comment:
The `zero` part of the name is a bit confusing. I think that documentation saying that it's 0 for fat32 is enough. Same for the other fields with `zero` in their name.